X509 app: major cleanup of user guidance, documentation, and code structure

DDvO opened the pull request:
While adding tests to #13658 I came across various shortcomings in the user guidance of the `x509` app.
Like with several other apps, there are meanwhile a lot of options and their presentation to users has become a mess.
This is likely due to various incremental additions of options having grown wild without intermediate cleanup.

I spent quite some effort getting the presentation of the options in the help output and the documentation into a logical order
and fixing descriptions where they were outdated or unclear.
During this process I inevitably found a couple of minor bugs and fixed them, 
in particular regarding unexpected combinations of options and memory leaks of the `-clrext` option.

~I've also renamed the new `-force_pubkey` option to the more consistent and less 'aggressive' name `-set_pubkey`~
and did some minor improvements to the related tests.

~I've renamed the `-signkey` option to `-key` for consistency with the `req` app
and because this better reflects that usually also the public portion is used,
retaining the old `-signkey` option for backward compatibility.~

Moreover, this adds the `-copy_extensions` option, used when transforming x509 <-> req.
Fixes #3638
Fixes #6481
Fixes #10458
Partly fixes #13708
Supersedes #9449

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

FdaSilvaYY commented:
typo in a commit message : `reasonalbe`

FdaSilvaYY commented on apps/x509.c:
```diff
@@ -705,6 +795,7 @@ int x509_main(int argc, char **argv)
                 BIO_puts(out, "\n");
             } else if (email == i || ocsp_uri == i) {
                 int j;
+
```
nit : this blank line is one line off

DDvO commented:
Thanks  @FdaSilvaYY for your comments. I've fixed the reported nits.

InfoHunter commented on apps/x509.c:
```diff
@@ -67,97 +69,117 @@ typedef enum OPTION_choice {
-    {"hash", OPT_HASH, '-', "Synonym for -subject_hash"},
+    {"startdate", OPT_STARTDATE, '-', "Print the notBefore field"},
+    {"enddate", OPT_ENDDATE, '-', "Print the notAfter field"},
+    {"dates", OPT_DATES, '-', "Print both Before and After dates"},
```
Should be `both notBefore and notAfter dates`?

DDvO replied:
Good point, so renamed them for consistency with the cert field names
and mentioned this in `CHANGES.md`.
I think it's not worth adding option aliases for backward compatibility, is it?

InfoHunter replied:
> I think it's not worth adding option aliases for backward compatibility, is it?

I think it's not worth, but I don't know if this would bring more inconvenient to the users.

And I think the text of option `OPT_DATES` should be `Print both notBefore and notAfter dates` since there are no fields called Before and After in an X.509 certificate ;-)

DDvO replied:
I've revised the help and documentation texts for all three options.

richsalz replied:
Having three options is kinda silly, unless you make notBefore and notAfter print out machine-readable, as in the ASN1 string?

DDvO replied:
> Having three options is kinda silly

I wonder if it's the right time to discuss such details on rather irrelevant options.
Let's focus now on getting the essence of the apps usable and well documented.

richsalz replied:
> I wonder if it's the right time to discuss such details on rather irrelevant options.
> Let's focus now on getting the essence of the apps usable and well documented.

I wonder if now is the right time to add new names to existing options.

mattcaswell commented:
> I've also renamed the new -force_pubkey option to the more consistent and less 'aggressive' name -set_pubkey

Just back from holiday and not had time to review this PR yet, but I noticed this in your description above. `-force_pubkey` is not new. It was introduced in 43206a2d7 back in 2011, and has been available in public releases since 1.0.2

DDvO commented:
> > I've also renamed the new -force_pubkey option to the more consistent and less 'aggressive' name -set_pubkey
> 
> Just back from holiday and not had time to review this PR yet, but I noticed this in your description above. `-force_pubkey` is not new. It was introduced in [43206a2](https://github.com/openssl/openssl/commit/43206a2d7cc87c959535c0f69e2aa3b364eafd6e) back in 2011, and has been available in public releases since 1.0.2

Oh right, I had confused the history of some options.
So I've reverted this change because this is not important and if done it would likely require an option alias for compatibility.

DDvO commented:
I've renamed the `-signkey` option to `-key` for consistency with the `req` app
and because this better reflects that usually also the public portion is used,
retaining the old `-signkey` option for backward compatibility.

DDvO commented:
Moreover, I've added the `-copy_extensions` option, used when transforming x509 <-> req.
Fixes #3638
Fixes #6481
Fixes #10458
Partly fixes #13708
Supersedes #9449

t8m commented on CHANGES.md:
```diff
@@ -231,6 +231,11 @@ OpenSSL 3.0
 
    *Richard Levitte*
 
+ * Renamed the `-startdate` option of the `x509` command to `-notBefore` and
```
Please drop this change. It is useless. The options were there since ages. And you're making this inconsistent with ca command.

DDvO replied:
> Please drop this change. It is useless. The options were there since ages. 
And you're making this inconsistent with ca command.

This change was suggested by @InfoHunter who has been with the project even longer than this option.
See the discussion above: https://github.com/openssl/openssl/pull/13711#pullrequestreview-556251087
To me, either variant is fine; does anyone else have a strong opinion on it?

t8m replied:
As I understand the original @InfoHunter comment it was about the help text, not about the names of the options.

DDvO replied:
Argh, indeed I must have entirely misunderstood his comment!
Followed your interpretation.

t8m reviewed (changes requested):
Could you please squash the fixups and drop all the rename options commits? I do not think it is a good idea to rename the options since they have been there for ages.

t8m commented:
The -signkey rename could be submitted as separate PR to consider. However I am still not sure it is worth it.

DDvO commented:
> The -signkey rename could be submitted as separate PR to consider. However I am still not sure it is worth it.

Before doing that, I'd like to learn if it's worth the effort of separating this change (and later merging with with conflicts).
This has been my motivation for the change:

> I've renamed the `-signkey` option to `-key` for consistency with the `req` app
and because this better reflects that usually also the public portion is used,
retaining the old `-signkey` option for backward compatibility.

What do others think about it?

DDvO commented:
> Could you please squash the fixups

Done, also rebased to the latest master.

t8m commented:
> Before doing that, I'd like to learn if it's worth the effort of separating this change (and later merging with with conflicts).
> This has been my motivation for the change:

You could just save the commit as is for later (as it is the last one in the PR anyway) and submit it only after this PR is merged.

Adding multiple unrelated things into a single PR makes it hard to review and hard to approve if someone disagrees some of the unrelated parts.

Please note that I do not completely disagree with the -signkey rename although I would keep the new option just as an alias and would not deprecate the old one as it is probably widely used already in various scripts and it is not worth removing the compatibility even any time later. So I would not block merging such rename. But I am less keen on explicitly approving it.

DDvO commented:
> > Before doing that, I'd like to learn if it's worth the effort of separating this change (and later merging with with conflicts).
> > This has been my motivation for the change:
> 
> You could just save the commit as is for later (as it is the last one in the PR anyway) and submit it only after this PR is merged.

That's a nice idea - will do that.
And I've taken that commit out of this PR as you advised.

> Adding multiple unrelated things into a single PR makes it hard to review and hard to approve if someone disagrees some of the unrelated parts.

Good point.

> Please note that I do not completely disagree with the -signkey rename although I would keep the new option just as an alias and would not deprecate the old one as it is probably widely used already in various scripts and it is not worth removing the compatibility even any time later. So I would not block merging such rename. But I am less keen on explicitly approving it.

I understand, so will adapt that PR to add an alias rather than deprecating the old option name.

DDvO commented:
@t8m, your recent change requests are fulfilled.

DDvO commented:
I've added a further related commit regarding `apps/x509.c`: 
Make `-x509toreq` respect `-clrext`, `-sigopt`, -and `-extfile` options
and prevent copying SKID and AKID extensions to CSRs, which make no sense there.

BTW, it's a historical accident that SKIDs and AKIDs are encoded as X.509 extensions rather than as genuine cert attributes.

richsalz commented:
> BTW, it's a historical accident that SKIDs and AKIDs are encoded as X.509 extensions rather than as genuine cert attributes.

ITU didn't define it, the IETF did.  i don't think anyone would say X509 is "minimal" but it really just has only what is strictly needed ... :)

t8m commented on doc/man1/openssl-x509.pod.in:
```diff
@@ -396,21 +398,20 @@ generate a certificate containing any desired public key.
 =item B<-clrext>
 
 When a transforming a certificate to a new certificate
```
When you're editing here please s/a transforming/transforming/

DDvO replied:
Done.

DDvO commented:
I've further improved my most recent commit such that now the `-ext` option can be used to select which extensions to copy.
On this occasion I further simplified the overall structure of the code, removing major redundancies.

DDvO commented:
BTW, my two most recent commits also fix memory leaks of the `-clrext` option, 
which surfaced in tests after I had used the same `X509_delete_ext(x, i)` pattern for `-ext`.

So I fixed both usages and added a warning to the doc of `X509v3_delete_ext()` and friends 
that their behavior (namely, not to deallocate their subject) is a bit counter-intuitive.

t8m commented on apps/x509.c:
```diff
@@ -58,138 +51,233 @@ typedef enum OPTION_choice {
+
+    if (OBJ_cmp(obj, excluded) == 0) {
+        if (names != NULL && strstr(names, sn) != NULL)
+            BIO_printf(bio_err,
+                       "Warning: refusing to copy %s extension specified in -ext option\n",
+                       sn);
```
I do not think we need this warning. It makes the command unnecessarily noisy because the skid and akid is present in most of the regular certificates produced by CAs.

DDvO replied:
In my opinion it is good to have this warning.
Note that it is shown only if `-x509toreq` and `-copy_extensions copy` is given and then still it is not shown by default,
but only when also the `-ext` option is given explicitly requesting to copy `subjectKeyIdentifier` or `authorityKeyIdentifier` to the cert request, which simply makes no sense and for this reason is refused.
So this warning is only given on a nonsensical explicit request (and thus could even be made an error).

t8m replied:
OK, I misunderstood the check slightly - if the warning is given only in case skid/akid is listed in -ext, then it is definitely ok, although it might be even better to just warn if skid/akid is listed in -ext without even checking that it is present in the certificate. So it is clear right away that trying to copy akid/skid won't work.

DDvO replied:
All right, adapted the warning this way.

t8m reviewed (changes requested):
A few nits to be fixed, otherwise looks good.

t8m commented:
And yeah, we need a rebase to fix conflicts.

DDvO commented on crypto/x509/x509_req.c:
```diff
@@ -236,6 +236,10 @@ int X509_REQ_add1_attr_by_NID(X509_REQ *req,
+	if (req == NULL || bytes == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_PASSED_NULL_PARAMETER);
+		return 0;
+	}
```
Oops, fixed.

DDvO commented on CHANGES.md:
```diff
@@ -900,6 +900,12 @@ OpenSSL 3.0
 
    *Richard Levitte*
 
+ * Added the `<-copy_extensions` option to the `x509` command for use with
```
Fixed, also a further such occurrence in the following entry.
On this occasion I've also improved the description in `CHANGES.m` on the new default SKID and AKID behavior.

DDvO commented on apps/x509.c:
```diff
@@ -58,138 +51,233 @@ typedef enum OPTION_choice {
+
+    if (sk_X509_EXTENSION_num(exts) > 0) {
+        if (ext_copy == EXT_COPY_UNSET) {
+            BIO_printf(bio_err,
```
I've dropped it as you wrote, though it would have been an automated way of making users aware of the new feature.
I believe users would expect that extensions are copied by default (as is done when converting one cert to another),
see all the requests for the extension copy feature listed here: #10458.

DDvO commented on apps/x509.c:
```diff
@@ -574,64 +690,69 @@ int x509_main(int argc, char **argv)
+        }
+        if (req != NULL) {
+            if (ext_copy == EXT_COPY_UNSET) {
+                BIO_printf(bio_err,
```
Done; see above.

DDvO commented on doc/man1/openssl-x509.pod.in:
```diff
@@ -98,356 +100,384 @@ various sections.
+
+=item B<-x509toreq>
+
+Output a PCSK#10 certificate request (rather than a certificate).
```
Fixed

DDvO commented on doc/man1/openssl-x509.pod.in:
```diff
@@ -803,12 +846,9 @@ must be present.
 It is possible to produce invalid certificates or requests by specifying the
-wrong private key or using inconsistent options in some cases: these should
-be checked.
+wrong private key, using unsuitable X.509 exteions,
```
Fixed

DDvO commented:
> A few nits to be fixed, otherwise looks good.

> And yeah, we need a rebase to fix conflicts.

Thanks @t8m for your further round of comments; 
I've addressed them and also fixed the recent merge conflicts.

t8m commented on apps/x509.c:
```diff
@@ -58,138 +51,230 @@ typedef enum OPTION_choice {
+     * If names is not NULL, copy only those extensions listed there.
+     */
+    if ((exts = sk_X509_EXTENSION_new_reserve(NULL, n)) == NULL)
+        return NULL;
```
Leak of req here.

DDvO replied:
Good spotting - fixed.

t8m commented on crypto/x509/x509_req.c:
```diff
@@ -236,6 +236,10 @@ int X509_REQ_add1_attr_by_NID(X509_REQ *req,
                               int nid, int type,
                               const unsigned char *bytes, int len)
 {
+	if (req == NULL || bytes == NULL) {
+            ERR_raise(ERR_LIB_X509, ERR_R_PASSED_NULL_PARAMETER);
+            return 0;
+	}
```
The indentation is still wrong for the whole if here.

DDvO replied:
Oops, fixed, and simplified the function.

t8m reviewed (changes requested):
Still some nits to solve.

DDvO commented:
> Still some nits to solve.

Done.

t8m commented on crypto/x509/x509_req.c:
```diff
@@ -236,14 +236,12 @@ int X509_REQ_add1_attr_by_NID(X509_REQ *req,
-                                type, bytes, len))
-        return 1;
-    return 0;
+    if (req == NULL || bytes == NULL) {
```
Now looking at this particular change in broader context it looks very awkward to me to check for the NULL parameters in just this particular function but not to do it in the other similar functions around. However I am not sure it is really worth it to add the checks as we do not do that universally throughout the code - what was the motivation for this particular change?

I'd recommend dropping it from this PR and if you want to, you could submit a different PR adding the NULL checks everywhere across these functions. But perhaps it should be discussed first on openssl-project ML whether we want such checks universally.

DDvO replied:
As far as i recall, my motivation for adding this check was that during the changes I did to `x509.c` I made a mistake that lead to the `req` argument to be NULL and thus to a crash.
I had learned during the CMP contribution that every public function should validate its input, and so it would certainly be an improvement to do it here. 

Yet I see your point that it appears imbalanced to not do this consistently for related functions. So following your recommendation I've dropped that change from this PR,
also because it is very bulky already anyway.

On this occasion I've squashed the remaining fixup commits.

t8m reviewed (approved):
This one was very hard to review so for the next time I'd suggest avoiding sideline improvements if possible. Submit them as separate PRs to keep the main topic PR more manageable.

Anyway this is really a good improvement of the code and it hopefully should not cause any serious regression.

DDvO commented:
> This one was very hard to review so for the next time I'd suggest avoiding sideline improvements if possible. Submit them as separate PRs to keep the main topic PR more manageable.

Yeah, unfortunately this PR became very bulky.

It started out by improving a few options of `x509` and their documentation,
but doing so I came across more and more issues related to these changes,
and I was forced to do some further fixes also in the crypto lib to get the tests through etc.

A couple of fixes could have been done independently, though.
The later addition of the `-copy_extensions` option would better have been done in a follow-up PR after this one would have been merged, but adding the same option to the `req` app I did not resist the temptation to add it here at the same time.

On the bright side, handling this all in one PR saved us from handling some 10+ separate PRs :wink:

DDvO commented:
If I had known in advance how much technical debt this PR would uncover and need to deal with I'd likely not have done it because it was a huge effort, which I ended up doing in my spare time.

I appreciate your endurance also as a reviewer.

> Anyway this is really a good improvement of the code 

Pleased to hear!

>and it hopefully should not cause any serious regression.

I hope so, too.
(And if there is one, this would mean there was a serious omission in the existing tests).

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Merged - thanks a lot @t8m for your effort reviewing this large PR!

kousu commented:
Thank you thank you thank you

hablutzel1 commented:
The following problem has been apparently introduced by the current PR around . 

If you create `cert.pem`:

```
-----BEGIN CERTIFICATE-----
MIIEZTCCA02gAwIBAgIQQAF1BIMUpMghjISpDBbN3zANBgkqhkiG9w0BAQsFADA/
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT
DkRTVCBSb290IENBIFgzMB4XDTIwMTAwNzE5MjE0MFoXDTIxMDkyOTE5MjE0MFow
MjELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUxldCdzIEVuY3J5cHQxCzAJBgNVBAMT
AlIzMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAuwIVKMz2oJTTDxLs
jVWSw/iC8ZmmekKIp10mqrUrucVMsa+Oa/l1yKPXD0eUFFU1V4yeqKI5GfWCPEKp
Tm71O8Mu243AsFzzWTjn7c9p8FoLG77AlCQlh/o3cbMT5xys4Zvv2+Q7RVJFlqnB
U840yFLuta7tj95gcOKlVKu2bQ6XpUA0ayvTvGbrZjR8+muLj1cpmfgwF126cm/7
gcWt0oZYPRfH5wm78Sv3htzB2nFd1EbjzK0lwYi8YGd1ZrPxGPeiXOZT/zqItkel
/xMY6pgJdz+dU/nPAeX1pnAXFK9jpP+Zs5Od3FOnBv5IhR2haa4ldbsTzFID9e1R
oYvbFQIDAQABo4IBaDCCAWQwEgYDVR0TAQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8E
BAMCAYYwSwYIKwYBBQUHAQEEPzA9MDsGCCsGAQUFBzAChi9odHRwOi8vYXBwcy5p
ZGVudHJ1c3QuY29tL3Jvb3RzL2RzdHJvb3RjYXgzLnA3YzAfBgNVHSMEGDAWgBTE
p7Gkeyxx+tvhS5B1/8QVYIWJEDBUBgNVHSAETTBLMAgGBmeBDAECATA/BgsrBgEE
AYLfEwEBATAwMC4GCCsGAQUFBwIBFiJodHRwOi8vY3BzLnJvb3QteDEubGV0c2Vu
Y3J5cHQub3JnMDwGA1UdHwQ1MDMwMaAvoC2GK2h0dHA6Ly9jcmwuaWRlbnRydXN0
LmNvbS9EU1RST09UQ0FYM0NSTC5jcmwwHQYDVR0OBBYEFBQusxe3WFbLrlAJQOYf
r52LFMLGMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjANBgkqhkiG9w0B
AQsFAAOCAQEA2UzgyfWEiDcx27sT4rP8i2tiEmxYt0l+PAK3qB8oYevO4C5z70kH
ejWEHx2taPDY/laBL21/WKZuNTYQHHPD5b1tXgHXbnL7KqC401dk5VvCadTQsvd8
S8MXjohyc9z9/G2948kLjmE6Flh9dDYrVYA9x2O+hEPGOaEOa1eePynBgPayvUfL
qjBstzLhWVQLGAkXXmNs+5ZnPBxzDJOLxhF2JIbeQAcH5H0tZrUlo5ZYyOqA7s9p
O5b85o3AM/OJ+CktFBQtfvBhcJVd9wvlwPsk+uyOy2HI7mNxKKgsBTt375teA2Tw
UdHkhVNcsAKX1H7GNNLOEADksd86wuoXvg==
-----END CERTIFICATE-----
```

And execute:

`openssl x509 -in cert.pem -issuer -noout`

Instead of getting the correct issuer (which is something like this):

`issuer= /O=Digital Signature Trust Co./CN=DST Root CA X3`

You get the subject: 

`issuer=C = US, O = Let's Encrypt, CN = R3`

PS: Sorry for such a naive bug report without a patch or nothing. I'm just out of time.

DDvO commented:
Thanks @hablutzel1 for your bug report. 
This is now being fixed in #14340.

DDvO commented:
> This is now being fixed in #14340.

Done

hablutzel1 commented:
Kind of off-topic here, but could you please point me to some official documentation on the OpenSSL development guidelines/practices (if it exists)?, as I have the impression that this PR includes commits that are not as granular as possible and I find it kind of dangerous as that could make commits harder to be appropriately reviewed which later could lead to the introduction of bugs... even security ones.

DDvO commented:
> Kind of off-topic here, but could you please point me to some official documentation on the OpenSSL development guidelines/practices (if it exists)?

There is a contributors guide, which is mentioned whenever a PR is being submitted: 
https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md
and is part of the code base, 
but for some reason is not referenced from https://www.openssl.org/policies/.

lzap commented:
For googlers, this is gonna be part of the **OpenSSL 3.0.0** release.

_(This is a longer wait than for my Playstation 5, anyway, thanks all :-)_

stas-sbi commented:
I don't know if this was an intent (I guess it wasn't) but this PR made ([this](https://github.com/openssl/openssl/commit/05458fdb73dcca30edace5ad727a15d6d919e215#diff-0fb19438117c9272e0d4a812eab3dfd079ad93e5e598fda46a6f2eb65fa9ad16L668-L669) change specifically) made `-CAcreateserial` and `-CAserial` options ignored. There is no serial file created and the serial is always random now.

giner commented:
I don't know if this was an intent (I guess it wasn't) but this PR made ([this](https://github.com/openssl/openssl/commit/05458fdb73dcca30edace5ad727a15d6d919e215#diff-0fb19438117c9272e0d4a812eab3dfd079ad93e5e598fda46a6f2eb65fa9ad16L668-L669) change specifically) made `-CAcreateserial` and `-CAserial` options ignored. There is no serial file created and the serial is always random now.

DDvO commented:
> I don't know if this was an intent (I guess it wasn't) but this PR ([this](https://github.com/openssl/openssl/commit/05458fdb73dcca30edace5ad727a15d6d919e215#diff-0fb19438117c9272e0d4a812eab3dfd079ad93e5e598fda46a6f2eb65fa9ad16L668-L669) change specifically) made `-CAcreateserial` and `-CAserial` options ignored. There is no serial file created and the serial is always random now.

Thanks for your comment.
I've just investigated: there is indeed a regression due to a mistake of mine in commit 05458fdb73dcca30edace5ad727a15d6d919e215 in case the `-new` option is given along with `-CA`.
I'll provide a fix for this issue (and a related one).

DDvO commented:
Fix is in #18373.
